File: Binary Search/162. Find Peak Element/162. Find Peak Element.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var findPeakElement = function (nums) {
  let low = 0;
  let high = nums.length;

  while (low <= high) {
    const mid = Math.floor((low + high) / 2);
    const left = mid === 0 || nums[mid] > nums[mid - 1];
    const right = mid === nums.length - 1 || nums[mid] > nums[mid + 1];
    if (left && right) return mid;
    else if (left) low = mid + 1;
    else high = mid - 1;
  }
};

// Time: O(log n)
// Space: O(1)

File: Binary Search/162. Find Peak Element/162. Find Peak Element.py
class Solution:
    def findPeakElement(self, nums: List[int]) -> int:
        low = 0
        high = len(nums) - 1

        while (low <= high):
            mid = (low + high) // 2
            left = mid == 0 or nums[mid] > nums[mid - 1]
            right = mid == len(nums) - 1 or nums[mid] > nums[mid + 1]
            if left and right:
                return mid
            elif left:
                low = mid + 1
            else:
                high = mid - 1
        
# Time: O(log n)
# Space: O(1)